### PR TITLE
Tim Zulf- MBP- In Progress

### PR DESCRIPTION
**Pull request by Tulf**
## Running the Project

- Fork and Clone this project.
- `cd` into your project folder.
- Run `npm install` or `yarn` to download the dependencies.
- Add `knex` and `sqlite3` npm modules.
- Configure `knex` to connect to `/data/lambda.sqlite3` using the `sqlite3` module.
- Write a set of endpionts inside `index.js` to satisfy the specifications listed below.
- To start the API server, run `yarn start` or `npm start`.
- Use _Postman_ to test your API.

### Table

The included database has a _zoos_ table with the following schema:

- id: integer, primary key, autoincrements.
- name: text, required, unique.

### `POST /api/zoos`

When the client makes a `POST` request to this endpoint, a new _zoo_ should be created in the _zoos_ table.

Ensure the client passes a `name` property in the request body. If there's an error, respond with an appropriate status code, and send a JSON response of the form `{ error: "Some useful error message" }`.

Return de `id` of the inserted zoo and a 201 status code.

### `GET /api/zoos`

When the client makes a `GET` request to this endpoint, return a list of all the _zoos_ in the database. Remember to handle any errors and return the correct status code.

### `GET /api/zoos/:id`

When the client makes a `GET` request to `/api/zoos/:id`, find the _zoo_ associated with the given `id`. Remember to handle errors and send the correct status code.

### DELETE /api/zoos/:id

When the client makes a `DELETE` request to this endpoint, the _zoo_ that has the provided `id` should be removed from the database.

### PUT /api/zoos/:id

When the client makes a `PUT` request to this endpoint passing an object with the changes, the _zoo_ with the provided `id` should be updated with the new information.

## Stretch Problem

Add a new _bears_ table to the database and add endpoints to perform CRUD operations on it. Each bear should have an `id` and `name` property similar to the _zoos_ table.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -8,6 +13,93 @@ server.use(helmet());
...
+    res.status(500).json(err)
+  })
+})
+//BY ID:
+
+server.get('/api/zoos/:id', (req, res) => {
+  const  {id} = req.params;
+  db('zoos')
+  .select()
+  .where('id', id)
```
Comment (Mister-Corn) [inline index.js]: I've seen like three different ways you can use `.where()`. Knex can be confusing. 😕

File: index.js [hunk 2]
```diff
@@ -8,6 +13,93 @@ server.use(helmet());
...
+})
+
+//-------------PUT-----------//
+server.put('/api/zoos/:id', (req, res) => {
+  const { id } = req.params;
+  const name = req.body;
+
+  db('zoos')
+  .where( { id })
+  .update(name)
```
Comment (Mister-Corn) [inline index.js]: Does the order of the chaining matter? I honestly do not know.

File: index.js [hunk 3]
```diff
@@ -8,6 +13,93 @@ server.use(helmet());
...
+    console.log(err)
+    res.status(500).json(err)
+  })
+})
+
+//----POST ------//
+
+server.post( '/api/zoos', (req, res) => {
+  const zoo = req.body;
+
```
Comment (Mister-Corn) [inline index.js]: This is perfectly fine, but try to do some validation/error-checking of the client input sometime.

**Comment by Mister-Corn** [comment]
You pushed your `node-modules` to Github. 😅 I will be going over `.gitignore` during tomorrow's (9/12/2018) stand-up.

Otherwise, good work! 👍

File: index.js (hunk 3 above)
Comment (Tulf) [reply index.js]: Yeah I was a little bit lazy, I spent a lot of time reading the knex documentation but it wasnt' as fruitful as I hoped so I ended up doing everything as fast as possible in the only way I understood it. And yeah going over gitignore would be helpful, I guess that's what the package.json is for after all haha.

Comment (Mister-Corn) [reply index.js]: lol good developers are ~lazy~ _efficient_. 😄 so no worries! it's just something to think about in the future, that's all.

**Pull request closed without merging**